Add `queryStakePoolDefaultVote` state query

Lucsanszky opened the pull request:
# Description
Resolves #4663 
<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

Lucsanszky committed:
Add `queryStakePoolDefaultVote` state query

This retrieves the delegatee of a stake pool's reward account,
which determines the default vote of the stake pool.

For clarification and discussion, see: #4645

lehins commented on libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs:
```diff
@@ -351,3 +357,21 @@ finishedPulserState ::
+-- | Query a stake pool's reward account delegatee which determines the pool's default vote
+-- in absence of an explicit vote. Note that this is different from the delegatee determined
+-- by the credential of the stake pool itself.
+querySPODefaultVote ::
```
We call them SPOs in tests and during conversations for the sake of brevity, but  we should not be using anywhere else ledger, since SPO is a person or an organization who **O**perates" a Stake Pool. Everything in ledger in reality is about stake pools, not stake pool operators. So, we should use StakePool in all publicly visible interface, instead of SPO. This point about the name is only about accuracy, but also about consistency.
```suggestion
queryStakePoolDefaultVote ::
```

Lucsanszky replied:
Good point! I'm familiar with the abbreviation but I had the impression that we prefer using SPO (which I never quite liked for the same reasons you point out). It seems like my impression was a false illusion stemming from the fact that I spent a lot of time in tests recently. :D

lehins commented on libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs:
```diff
@@ -351,3 +357,21 @@ finishedPulserState ::
+  --       otherwise @No@
+  --     * Other (@Nothin@ or @DRepKeyHash@, @DRepScriptHash@): default vote is @No@
+  -- In case of a @HardForkInitiation@, the default vote is always @No@.
+  Maybe (DRep (EraCrypto era))
```
We need to create a default vote type, instead of returning a DRep, because obtaining a DRep that StakePool's reward account delegated to, does not mean much, unless it is a predefined DRep. So, we should make it for the user absolutely clear through a specialized type what is going on.
```suggestion
  DefaultVote
```
where
```haskell
data DefaultVote =
  = DefaultNo
  | DefaultAbstain
  | DefaultNoConfidence
```

Lucsanszky replied:
Sounds good! :+1: I was thinking about this as well but since I was not sure if we want to do it, I opted to provide an explanation in the docs as a first draft instead. I'll make the change.

Lucsanszky replied:
I added the data type in this module and exported it in case someone downstream wants to use it. Let me know if you want the type to be defined somewhere else or if you think we shouldn't expose the data type to users.

lehins reviewed (commented):
Looks good. Couple of improvements suggested.

lehins commented on libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs:
```diff
@@ -351,3 +358,32 @@ finishedPulserState ::
+    DefaultAbstain
+  | -- | Reward account is delegated to a @DRepKeyHash@, @DRepScriptHash@ or undelegated:
+    --   default vote is @No@.
+    DefaultNoConfidence
```
I think the order of haddock got a little screwed up.
```suggestion
  = -- | Reward account is delegated to a @DRepKeyHash@, @DRepScriptHash@ or undelegated:
    --   default vote is @No@.
    DefaultNo
  | -- | Reward account is delegated to @DRepAlwaysAbstain@:
    --   default vote is @Abstain@, except for @HardForkInitiation@ actions. 
    DefaultAbstain
  | -- | Reward account is delegated to @DRepAlwaysNoConfidence@:
    --   default vote is @Yes@ in case of a @NoConfidence@ action, otherwise @No@.
    DefaultNoConfidence
```

lehins reviewed (approved):
Looks great. Thank you!

lehins committed:
Reduce duplication by extracting common functionality

RATIFY rule and ledger state query have very similar functionality. For
safety and reduction of duplication it was possible to extract this
logic into a separate function `defaultStakePoolVote` and reuse it in
both places.

lehins reviewed (approved):
@Lucsanszky I've added one more commit that reduces duplication and reuses the same function in RATIFY rule as well as in the ledger state query.

Lucsanszky commented:
> @Lucsanszky I've added one more commit that reduces duplication and reuses the same function in RATIFY rule as well as in the ledger state query.

Great addition, thanks!
